### PR TITLE
Dehardcoded skill and damage requirements in martialarts.cpp

### PR DESCRIPTION
**Pull request by Hymore246**
#### Summary

```SUMMARY: Content "Dehardcoded skill and damage requirements in martialarts.cpp"```

#### Purpose of change

Any skill or damage type can be used as a requirement for martial arts buffs and techniques.

#### Describe the solution

Previously, only Melee, Unarmed Combat, Cutting Weapons, Stabbing Weapons, or Bashing Weapons skills could be used as skill requirements for martial art techniques and buffs. Likewise, only Bash, Cut, or Stab were valid weapon damage types for techniques and buffs. With this change, skills are now loaded dynamically from the Skill::skill list and damage types from the dt_map list from when a martial art is generated. If the skill or damage type exists in the game, it can be used as a martial arts requirement. The JSON format to do this is:

```
"skill_requirements": [ { "name": "<ident>", "level": <skill_level> } ]
```
Where "ident" is the JSON ident field of the skill in question. Unarmed combat would be "unarmed", Dodging would be "dodge", etc.

Minimum Damage Type is a seldomly used weapon requirement in the martial arts system. It checks to see if the weapon the player is using does at least x amount of a certain damage type. Only the base weapon damage matters, everything else is ignored. Min Damage doesn't show up as a requirement in descriptions so this section of the martial arts system is rarely seen. The switch to dynamically loading from the game's damage map (dt_map) made it so the old version of this attributes would need to be updated. Thankfully, the only technique that used it was the old version was Eskrima's Puno Strike, so nothing had to be changed. The JSON format to do this is:

```
"weapon_damage_requirements": [ { "type": "<damage type>", "min": <minimum damage> } ]
```

Where "damage" is the type of damage (bash, heat, cut, acid, etc) that the weapon needs. 

For mod makers, you might be wondering: "Is it possible to use a new skill I added in my mod as a requirement for a new martial art?"

The answer is YES.

Because the skills are being read by a JSON reader object, you can add custom skill requirements to any buff or technique (both old and mod specific). However, adding a skill requirement that does not exist will create a skill requirement for "missing" which will keep the buff/technique from triggering.

#### Additional context

Below is an example of the new dynamic skill requirements.
new_ma_skills

MAGICAL Panzer Kunst is a *totally* legit style that I didn't added to Magiclysm by copy, pasting, and editing files from the MMA mod.

### CODE REVIEW

**Comment by foulman** [comment]
Magical Martial Arts, here we come!

**Comment by cainiaowu** [comment]
This json format kind of feels weird to me, since it is mixing key with value.
Why not use 
```
"min_skill" :<ident>
"min_skill_level": <skill level>
```

**Comment by Hymore246** [comment]
> 
> 
> This json format kind of feels weird to me, since it is mixing key with value.
> Why not use
> 
> ```
> "min_skill" :<ident>
> "min_skill_level": <skill level>
> ```

Because it's possible to use multiple skills at different levels as prerequisites:

"min_unarmed": 3
"min_melee": 1

vs

"min_skill": "unarmed",
"min_skill_level": 1
"min_skill": "melee",
"min_skill_level": 3

The above would not work because you can't have more than one of the same attribute in JSON. That throws a big error and kicks you out of loading CDDA.

**Comment by Qrox** [comment]
You can alternatively use JSON arrays to store that information, for example

```
"min_skill_and_level": [ [ "unarmed", 1 ], [ "melee", 3 ] ],
```

File: src/martialarts.cpp [hunk 1]
```diff
@@ -84,16 +84,24 @@ void ma_requirements::load( JsonObject &jo, const std::string & )
...
-    add_if_exists( jo, min_skill, was_loaded, "min_cutting", skill_cutting );
-    add_if_exists( jo, min_skill, was_loaded, "min_stabbing", skill_stabbing );
-
-    add_if_exists( jo, min_damage, was_loaded, "min_bashing_damage", DT_BASH );
-    add_if_exists( jo, min_damage, was_loaded, "min_cutting_damage", DT_CUT );
-    add_if_exists( jo, min_damage, was_loaded, "min_stabbing_damage", DT_STAB );
+    // Minimum skill level attributes
+    for( const auto &s : Skill::skills ) {
+        std::string skill_name = s.ident().str();
+        std::string attribute_name = string_format( _( "min_%s" ), skill_name );
```
Comment (Qrox) [inline src/martialarts.cpp]:
```suggestion
        std::string attribute_name = string_format( "min_%s", skill_name );
```

This shouldn't be translated.

File: src/martialarts.cpp [hunk 2]
```diff
@@ -84,16 +84,24 @@ void ma_requirements::load( JsonObject &jo, const std::string & )
...
+
+        add_if_exists( jo, min_skill, was_loaded, attribute_name, s.ident() );
+    }
+
+
+    // Minimum damage type attributes
+    std::map<std::string, damage_type> dt_map = get_dt_map();
+    auto iter = dt_map.cbegin();
+    while( iter != dt_map.cend() ) {
+        std::string attribute_name = string_format( _( "min_%s_damage" ), iter->first );
```
Comment (Qrox) [inline src/martialarts.cpp]:
```suggestion
        std::string attribute_name = string_format( "min_%s_damage", iter->first );
```

This shouldn't be translated.

**Comment by Hymore246** [comment]
> 
> 
> You can alternatively use JSON arrays to store that information, for example
> 
> ```
> "min_skill_and_level": [ [ "unarmed", 1 ], [ "melee", 3 ] ],
> ```

That format is much better. It probably should be something like that but it's going to be a lot of work to change the JSON and C++ code to get that to work. Something to work on in the future.

**Comment by kevingranade** [comment]
How about:
```
"skill_requirements": [
  { "skill": "unarmed", "min": 5, "max": 9 },
  { "skill": "magic", "min": 2 }
]
```
I'm REALLY not a fan of parsing things out of keys, to the point where I'm not sure having them hardcoded is worse.

**Comment by Hymore246** [comment]
> 
> 
> How about:
> 
> ```
> "skill_requirements": [
>   { "skill": "unarmed", "min": 5, "max": 9 },
>   { "skill": "magic", "min": 2 }
> ]
> ```
> 
> I'm REALLY not a fan of parsing things out of keys, to the point where I'm not sure having them hardcoded is worse.

I think I can used the skills reader from profession.cpp to do this. It will probably end up looking like this:
```
"skill_requirements": [
  { name: "melee", "level": 5 },
  { name: "unarmed", "level": 1 },
  { name: "spellcraft", "level": 7 }
]
```

I will need to go back and update martialarts.json and techniques.json with the new format. I'm not sure if I should do it now or wait till the rebalance completes.

**Comment by Hymore246** [comment]
Should be good now assuming the checks pass.

**Comment by Qrox** [comment]
Looks like your merge resulted in some recent changes being reverted.

**Comment by Hymore246** [comment]
> 
> 
> Looks like your merge resulted in some recent changes being reverted.

Good catch. I'm not sure why it did that. I merged it with master and it didn't pick up those changes.

File: src/martialarts.cpp [hunk 3]
```diff
@@ -470,6 +491,18 @@ std::string ma_requirements::get_description( bool buff ) const
         }, enumeration_conjunction::none ) << std::endl;
     }
 
+    if( std::any_of( min_damage.begin(), min_damage.end(), []( const std::pair<damage_type, int> &pr ) {
+    return pr.second > 0;
+} ) ) {
+        dump << string_format( _( "<bold>Damage %s required: </bold>" ),
```
Comment (Qrox) [inline src/martialarts.cpp]:
```suggestion
        dump << ngettext( "<bold>Damage type required: </bold>",
            "<bold>Damage types required: </bold>", min_damage.size() );
```

File: src/martialarts.cpp [hunk 4]
```diff
@@ -470,6 +491,18 @@ std::string ma_requirements::get_description( bool buff ) const
...
+            ngettext( "type", "types", min_damage.size() ) );
```
Comment (Qrox) [inline src/martialarts.cpp]:
```suggestion
```

File: src/martialarts.cpp [hunk 5]
```diff
@@ -470,6 +491,18 @@ std::string ma_requirements::get_description( bool buff ) const
...
+
+        dump << enumerate_as_string( min_damage.begin(),
+        min_damage.end(), []( const std::pair<damage_type, int> &pr ) {
+            return string_format( "%s: <stat>%d</stat>", name_by_dt( pr.first ), pr.second );
```
Comment (Qrox) [inline src/martialarts.cpp]:
```suggestion
            return string_format( _( "%s: <stat>%d</stat>" ), name_by_dt( pr.first ), pr.second );
```

**Comment by Hymore246** [comment]
Alrighty, at this point I have no idea how to get martialarts.cpp to pass it's Astyle check. Travis's Astyle error amounts to "it's wrong" with no indication on what the problem is.  Mirroring existing code structure isn't working and running the code through an Astyle program isn't working either.

Any help or insight would be greatly appreciated.

**Comment by AMurkin** [comment]
Artistic Style Version 3.0.1

```diff
diff --git a/src/martialarts.cpp b/src/martialarts.cpp
index 315b219213..ee863ee170 100644
--- a/src/martialarts.cpp
+++ b/src/martialarts.cpp
@@ -491,11 +491,12 @@ std::string ma_requirements::get_description( bool buff ) const
         }, enumeration_conjunction::none ) << std::endl;
     }
 
-    if( std::any_of( min_damage.begin(), min_damage.end(), []( const std::pair<damage_type, int>  &pr ) {
+    if( std::any_of( min_damage.begin(),
+    min_damage.end(), []( const std::pair<damage_type, int>  &pr ) {
     return pr.second > 0;
 } ) ) {
         dump << ngettext( "<bold>Damage type required: </bold>",
-            "<bold>Damage types required: </bold>", min_damage.size() );
+                          "<bold>Damage types required: </bold>", min_damage.size() );
 
         dump << enumerate_as_string( min_damage.begin(),
         min_damage.end(), []( const std::pair<damage_type, int> &pr ) {
```

**Comment by Hymore246** [comment]
> Artistic Style Version 3.0.1

Looks like that did it. Thanks. I need to download a new Astyle tool.

File: src/martialarts.cpp [hunk 6]
```diff
@@ -74,6 +68,41 @@ void add_if_exists( JsonObject &jo, Container &cont, bool was_loaded,
...
+        std::map<std::string, damage_type> dt_map = get_dt_map();
+
+        std::pair<damage_type, int> get_next( JsonIn &jin ) const {
+            JsonObject jo = jin.get_object();
+            const damage_type dt = dt_map.find( jo.get_string( "type" ) )->second;
+            return std::pair<damage_type, int>( dt, jo.get_int( "min" ) );
+        }
+        template<typename C>
+        void erase_next( JsonIn &jin, C &container ) const {
+            damage_type id = dt_map.find( jin.get_string() )->second;
```
Comment (BevapDin) [inline src/martialarts.cpp]: You must check the result of `find` before using it. What happens when a modder accidentally writes a damage type that does not exist? A simple spelling mistake will crash the game.

File: src/martialarts.cpp [hunk 7]
```diff
@@ -91,12 +91,17 @@ class ma_weapon_damage_reader : public generic_typed_reader<ma_weapon_damage_rea
 
         std::pair<damage_type, int> get_next( JsonIn &jin ) const {
             JsonObject jo = jin.get_object();
-            const damage_type dt = dt_map.find( jo.get_string( "type" ) )->second;
+            std::string type = jo.get_string( "type" );
+            std::string s = dt_map.find( type ) != dt_map.end() ? type : "true";
```
Comment (Qrox) [inline src/martialarts.cpp]: It's better to throw a JSON error instead if `type` is not found in `dt_map`.

File: src/martialarts.cpp [hunk 8]
```diff
@@ -91,12 +91,17 @@ class ma_weapon_damage_reader : public generic_typed_reader<ma_weapon_damage_rea
...
+            const damage_type dt = dt_map.find( s )->second;
             return std::pair<damage_type, int>( dt, jo.get_int( "min" ) );
         }
         template<typename C>
         void erase_next( JsonIn &jin, C &container ) const {
-            damage_type id = dt_map.find( jin.get_string() )->second;
+            JsonObject jo = jin.get_object();
+            std::string type = jo.get_string( "type" );
+            std::string s = dt_map.find( type ) != dt_map.end() ? type : "true";
```
Comment (Qrox) [inline src/martialarts.cpp]: It's better to throw a JSON error instead if `type` is not found in `dt_map`.

Also, the code before the last commit was reading a string directly from `jin`, but the current code is reading a string with the key `"type"` inside an object. Is this intended?

File: src/damage.cpp [hunk 9]
```diff
@@ -237,6 +237,11 @@ static const std::map<std::string, damage_type> dt_map = {
     { translate_marker_context( "damage type", "electric" ), DT_ELECTRIC }
 };
 
+std::map<std::string, damage_type> get_dt_map()
```
Comment (BevapDin) [inline src/damage.cpp]:
```suggestion
const std::map<std::string, damage_type> &get_dt_map()
```
There is really no need to copy this.

File: src/martialarts.cpp [hunk 10]
```diff
@@ -74,6 +68,46 @@ void add_if_exists( JsonObject &jo, Container &cont, bool was_loaded,
...
+            reader_detail::handler<C>().erase_if( container, [&id]( const std::pair<skill_id, int> &e ) {
+                return e.first == id;
+            } );
+        }
+};
+
+class ma_weapon_damage_reader : public generic_typed_reader<ma_weapon_damage_reader>
+{
+    public:
+        std::map<std::string, damage_type> dt_map = get_dt_map();
```
Comment (BevapDin) [inline src/martialarts.cpp]: Why store a local copy of that map? The map will never change and the copy is not changed either, so why make a copy?

File: src/martialarts.cpp [hunk 11]
```diff
@@ -74,6 +68,46 @@ void add_if_exists( JsonObject &jo, Container &cont, bool was_loaded,
...
+            std::string s = dt_map.find( type ) != dt_map.end() ? type : "true";
+            const damage_type dt = dt_map.find( s )->second;
+            return std::pair<damage_type, int>( dt, jo.get_int( "min" ) );
+        }
+        template<typename C>
+        void erase_next( JsonIn &jin, C &container ) const {
+            JsonObject jo = jin.get_object();
+            std::string type = jo.get_string( "type" );
+            std::string s = dt_map.find( type ) != dt_map.end() ? type : "true";
+            damage_type id = dt_map.find( s )->second;
```
Comment (BevapDin) [inline src/martialarts.cpp]: Suggestion (only works when `get_dt_map` returns a reference):
```C++
const auto iter = get_dt_map().find( s );
if( iter == get_dt_map().end() ) {
    jin.throw_error("Invalid damage type");
}
const damage_type id = iter->second;
```
This avoids the repeated lookup in the map (the two calls to `find`).

File: src/martialarts.cpp [hunk 12]
```diff
@@ -74,6 +68,52 @@ void add_if_exists( JsonObject &jo, Container &cont, bool was_loaded,
...
+        std::map<std::string, damage_type> dt_map = get_dt_map();
+
+        std::pair<damage_type, int> get_next( JsonIn &jin ) const {
+            JsonObject jo = jin.get_object();
+            std::string type = jo.get_string( "type" );
+            const auto iter = get_dt_map().find( type );
+            if( iter == get_dt_map().end() ) {
+                jo.throw_error( "Invalid damage type" );
+            }
+            const damage_type dt = dt_map.find( type )->second;
```
Comment (BevapDin) [inline src/martialarts.cpp]:
```suggestion
            const damage_type dt = iter->second;
```

File: src/martialarts.cpp [hunk 13]
```diff
@@ -74,6 +68,52 @@ void add_if_exists( JsonObject &jo, Container &cont, bool was_loaded,
...
+        }
+        template<typename C>
+        void erase_next( JsonIn &jin, C &container ) const {
+            JsonObject jo = jin.get_object();
+            std::string type = jo.get_string( "type" );
+            const auto iter = get_dt_map().find( type );
+            if( iter == get_dt_map().end() ) {
+                jo.throw_error( "Invalid damage type" );
+            }
+            damage_type id = dt_map.find( type )->second;
```
Comment (BevapDin) [inline src/martialarts.cpp]:
```suggestion
            damage_type id = iter->second;
```

**Pull request merged**